File: api/middleware/validation.go
// api/middleware/validation.go
package middleware

import (
	"net/http"
	"regexp"
	"strings"

	"github.com/gin-gonic/gin"
)

// ValidationRule represents a validation rule
type ValidationRule struct {
	Required bool
	MinLen   int
	MaxLen   int
	Pattern  string
	Message  string
}

// ValidationRules defines validation rules for different fields
var ValidationRules = map[string]ValidationRule{
	"email": {
		Required: true,
		Pattern:  `^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$`,
		Message:  "Invalid email format",
	},
	"password": {
		Required: true,
		MinLen:   8,
		MaxLen:   128,
		Pattern:  `^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)[a-zA-Z\d@$!%*?&]{8,}$`,
		Message:  "Password must be at least 8 characters with uppercase, lowercase, and number",
	},
	"firstName": {
		Required: true,
		MinLen:   2,
		MaxLen:   50,
		Pattern:  `^[a-zA-ZáéíóúÁÉÍÓÚñÑ\s]+$`,
		Message:  "First name must be 2-50 characters and contain only letters",
	},
	"lastName": {
		Required: true,
		MinLen:   2,
		MaxLen:   50,
		Pattern:  `^[a-zA-ZáéíóúÁÉÍÓÚñÑ\s]+$`,
		Message:  "Last name must be 2-50 characters and contain only letters",
	},
	"phone": {
		Required: false,
		Pattern:  `^[\+]?[1-9][\d]{0,15}$`,
		Message:  "Invalid phone number format",
	},
	"title": {
		Required: true,
		MinLen:   3,
		MaxLen:   200,
		Message:  "Title must be 3-200 characters",
	},
	"description": {
		Required: true,
		MinLen:   10,
		MaxLen:   2000,
		Message:  "Description must be 10-2000 characters",
	},
	"role": {
		Required: true,
		Pattern:  `^(admin|office_manager|staff|counselor|psychologist|client)$`,
		Message:  "Invalid role. Must be one of: admin, office_manager, staff, counselor, psychologist, client",
	},
	"status": {
		Required: true,
		Pattern:  `^(open|in_progress|closed|pending|archived)$`,
		Message:  "Invalid status. Must be one of: open, in_progress, closed, pending, archived",
	},
	"priority": {
		Required: true,
		Pattern:  `^(low|medium|high)$`,
		Message:  "Invalid priority. Must be one of: low, medium, high",
	},
}

// ValidationError represents a validation error
type ValidationError struct {
	Field   string `json:"field"`
	Message string `json:"message"`
}

// ValidationResponse represents the validation response
type ValidationResponse struct {
	Valid   bool             `json:"valid"`
	Errors  []ValidationError `json:"errors"`
	Message string           `json:"message"`
}

// ValidateField validates a single field
func ValidateField(fieldName, value string, rule ValidationRule) *ValidationError {
	// Check required
	if rule.Required && strings.TrimSpace(value) == "" {
		return &ValidationError{
			Field:   fieldName,
			Message: rule.Message,
		}
	}

	// Skip other validations if field is empty and not required
	if strings.TrimSpace(value) == "" {
		return nil
	}

	// Check minimum length
	if rule.MinLen > 0 && len(value) < rule.MinLen {
		return &ValidationError{
			Field:   fieldName,
			Message: rule.Message,
		}
	}

	// Check maximum length
	if rule.MaxLen > 0 && len(value) > rule.MaxLen {
		return &ValidationError{
			Field:   fieldName,
			Message: rule.Message,
		}
	}

	// Check pattern
	if rule.Pattern != "" {
		matched, err := regexp.MatchString(rule.Pattern, value)
		if err != nil || !matched {
			return &ValidationError{
				Field:   fieldName,
				Message: rule.Message,
			}
		}
	}

	return nil
}

// ValidateRequest validates a request against specified rules
func ValidateRequest(c *gin.Context, rules map[string]ValidationRule) *ValidationResponse {
	var errors []ValidationError

	for fieldName, rule := range rules {
		value := c.PostForm(fieldName)
		if value == "" {
			// Try to get from JSON body
			var jsonData map[string]interface{}
			if err := c.ShouldBindJSON(&jsonData); err == nil {
				if val, exists := jsonData[fieldName]; exists {
					if strVal, ok := val.(string); ok {
						value = strVal
					}
				}
			}
		}

		if err := ValidateField(fieldName, value, rule); err != nil {
			errors = append(errors, *err)
		}
	}

	return &ValidationResponse{
		Valid:   len(errors) == 0,
		Errors:  errors,
		Message: "Validation failed",
	}
}

// ValidationMiddleware creates a validation middleware
func ValidationMiddleware(rules map[string]ValidationRule) gin.HandlerFunc {
	return func(c *gin.Context) {
		validation := ValidateRequest(c, rules)
		
		if !validation.Valid {
			c.JSON(http.StatusBadRequest, gin.H{
				"error":   "Validation failed",
				"message": validation.Message,
				"details": validation.Errors,
			})
			c.Abort()
			return
		}

		c.Next()
	}
}

// Common validation middleware functions

// ValidateUserRegistration validates user registration data
func ValidateUserRegistration() gin.HandlerFunc {
	rules := map[string]ValidationRule{
		"firstName": ValidationRules["firstName"],
		"lastName":  ValidationRules["lastName"],
		"email":     ValidationRules["email"],
		"password":  ValidationRules["password"],
		"role":      ValidationRules["role"],
	}
	return ValidationMiddleware(rules)
}

// ValidateUserUpdate validates user update data
func ValidateUserUpdate() gin.HandlerFunc {
	rules := map[string]ValidationRule{
		"firstName": ValidationRules["firstName"],
		"lastName":  ValidationRules["lastName"],
		"email":     ValidationRules["email"],
		"phone":     ValidationRules["phone"],
		"role":      ValidationRules["role"],
	}
	return ValidationMiddleware(rules)
}

// ValidateCaseCreation validates case creation data
func ValidateCaseCreation() gin.HandlerFunc {
	rules := map[string]ValidationRule{
		"title":       ValidationRules["title"],
		"description": ValidationRules["description"],
		"category": {
			Required: true,
			MinLen:   2,
			MaxLen:   50,
			Message:  "Category must be 2-50 characters",
		},
		"priority": ValidationRules["priority"],
	}
	return ValidationMiddleware(rules)
}

// ValidateCaseUpdate validates case update data
func ValidateCaseUpdate() gin.HandlerFunc {
	rules := map[string]ValidationRule{
		"title":       ValidationRules["title"],
		"description": ValidationRules["description"],
		"status":      ValidationRules["status"],
		"priority":    ValidationRules["priority"],
	}
	return ValidationMiddleware(rules)
}

// ValidateAppointmentCreation validates appointment creation data
func ValidateAppointmentCreation() gin.HandlerFunc {
	rules := map[string]ValidationRule{
		"title": {
			Required: true,
			MinLen:   3,
			MaxLen:   200,
			Message:  "Title must be 3-200 characters",
		},
		"startTime": {
			Required: true,
			Pattern:  `^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(\.\d{3})?Z?$`,
			Message:  "Invalid start time format. Use ISO 8601 format",
		},
		"endTime": {
			Required: true,
			Pattern:  `^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(\.\d{3})?Z?$`,
			Message:  "Invalid end time format. Use ISO 8601 format",
		},
	}
	return ValidationMiddleware(rules)
}

// ValidateTaskCreation validates task creation data
func ValidateTaskCreation() gin.HandlerFunc {
	rules := map[string]ValidationRule{
		"title":       ValidationRules["title"],
		"description": ValidationRules["description"],
		"priority":    ValidationRules["priority"],
	}
	return ValidationMiddleware(rules)
}

// ValidateOfficeCreation validates office creation data
func ValidateOfficeCreation() gin.HandlerFunc {
	rules := map[string]ValidationRule{
		"name": {
			Required: true,
			MinLen:   2,
			MaxLen:   100,
			Message:  "Office name must be 2-100 characters",
		},
		"address": {
			Required: true,
			MinLen:   10,
			MaxLen:   200,
			Message:  "Address must be 10-200 characters",
		},
		"phone": ValidationRules["phone"],
		"email": ValidationRules["email"],
	}
	return ValidationMiddleware(rules)
}

// SanitizeInput sanitizes input to prevent XSS and other attacks
func SanitizeInput(input string) string {
	// Remove potentially dangerous characters
	input = strings.ReplaceAll(input, "<script", "")
	input = strings.ReplaceAll(input, "</script", "")
	input = strings.ReplaceAll(input, "javascript:", "")
	input = strings.ReplaceAll(input, "onload=", "")
	input = strings.ReplaceAll(input, "onerror=", "")
	
	// Trim whitespace
	input = strings.TrimSpace(input)
	
	return input
}

// SanitizeMiddleware sanitizes request data
func SanitizeMiddleware() gin.HandlerFunc {
	return func(c *gin.Context) {
		// Sanitize form data
		for key, values := range c.Request.PostForm {
			for i, value := range values {
				c.Request.PostForm[key][i] = SanitizeInput(value)
			}
		}

		c.Next()
	}
}

File: admin-portal/src/core/validation.ts
// admin-portal/src/core/validation.ts
import { ValidationError } from './errors';

// Validation Rules
export interface ValidationRule<T = any> {
  validate: (value: T) => boolean;
  message: string;
}

export interface ValidationSchema<T = any> {
  [key: string]: ValidationRule<T> | ValidationRule<T>[];
}

// Common Validation Rules
export const ValidationRules = {
  required: (message = 'This field is required'): ValidationRule => ({
    validate: (value: any) => value !== null && value !== undefined && value !== '',
    message,
  }),

  email: (message = 'Invalid email address'): ValidationRule<string> => ({
    validate: (value: string) => /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(value),
    message,
  }),

  minLength: (min: number, message?: string): ValidationRule<string> => ({
    validate: (value: string) => value.length >= min,
    message: message || `Must be at least ${min} characters`,
  }),

  maxLength: (max: number, message?: string): ValidationRule<string> => ({
    validate: (value: string) => value.length <= max,
    message: message || `Must be no more than ${max} characters`,
  }),

  pattern: (regex: RegExp, message = 'Invalid format'): ValidationRule<string> => ({
    validate: (value: string) => regex.test(value),
    message,
  }),

  oneOf: (options: any[], message?: string): ValidationRule => ({
    validate: (value: any) => options.includes(value),
    message: message || `Must be one of: ${options.join(', ')}`,
  }),

  date: (message = 'Invalid date'): ValidationRule<string> => ({
    validate: (value: string) => !isNaN(Date.parse(value)),
    message,
  }),

  futureDate: (message = 'Date must be in the future'): ValidationRule<string> => ({
    validate: (value: string) => new Date(value) > new Date(),
    message,
  }),
};

// Validator Class
export class Validator {
  static validateField<T>(value: T, rules: ValidationRule<T> | ValidationRule<T>[]): ValidationError | null {
    const ruleArray = Array.isArray(rules) ? rules : [rules];

    for (const rule of ruleArray) {
      if (!rule.validate(value)) {
        return new ValidationError(rule.message);
      }
    }

    return null;
  }

  static validateObject<T extends Record<string, any>>(
    data: T,
    schema: ValidationSchema
  ): Record<string, ValidationError> {
    const errors: Record<string, ValidationError> = {};

    for (const [field, rules] of Object.entries(schema)) {
      const value = data[field];
      const error = this.validateField(value, rules);

      if (error) {
        errors[field] = error;
      }
    }

    return errors;
  }

  static hasErrors(errors: Record<string, ValidationError>): boolean {
    return Object.keys(errors).length > 0;
  }

  static getFirstError(errors: Record<string, ValidationError>): ValidationError | null {
    const firstKey = Object.keys(errors)[0];
    return firstKey ? errors[firstKey] : null;
  }
}

// Common Validation Schemas
export const ValidationSchemas = {
  case: {
    title: [ValidationRules.required(), ValidationRules.minLength(3), ValidationRules.maxLength(200)],
    description: [ValidationRules.maxLength(2000)],
    client_id: ValidationRules.required('Client is required'),
    office_id: ValidationRules.required('Office is required'),
    category: ValidationRules.required('Category is required'),
    priority: ValidationRules.oneOf(['low', 'medium', 'high', 'urgent'], 'Invalid priority level'),
  },

  appointment: {
    title: [ValidationRules.required(), ValidationRules.minLength(3), ValidationRules.maxLength(200)],
    description: [ValidationRules.maxLength(1000)],
    case_id: ValidationRules.required('Case is required'),
    start_time: [ValidationRules.required('Start time is required'), ValidationRules.date()],
    end_time: [ValidationRules.required('End time is required'), ValidationRules.date()],
    status: ValidationRules.oneOf(['pending', 'confirmed', 'completed', 'cancelled', 'no_show']),
  },

  user: {
    email: [ValidationRules.required(), ValidationRules.email()],
    first_name: [ValidationRules.required(), ValidationRules.minLength(2), ValidationRules.maxLength(50)],
    last_name: [ValidationRules.required(), ValidationRules.minLength(2), ValidationRules.maxLength(50)],
    role: ValidationRules.oneOf([
      'admin', 'office_manager', 'lawyer', 'psychologist',
      'receptionist', 'event_coordinator', 'client'
    ]),
  },

  office: {
    name: [ValidationRules.required(), ValidationRules.minLength(2), ValidationRules.maxLength(100)],
  },
};

// Validation Helper Functions
export function validateCase(data: any) {
  return Validator.validateObject(data, ValidationSchemas.case);
}

export function validateAppointment(data: any) {
  return Validator.validateObject(data, ValidationSchemas.appointment);
}

export function validateUser(data: any) {
  return Validator.validateObject(data, ValidationSchemas.user);
}

export function validateOffice(data: any) {
  return Validator.validateObject(data, ValidationSchemas.office);
}
